HW0/P5/P5.txt

```diff
@@ -1 +1,86 @@
+
+(2) When we build a binary tree, we need d = ceil( ld (n) ) levels to compute the total. In this problem, we have a total of 8 employers who can work on this task. The first level consists of 256 bag, 8 employers can handle 16 bags in 1s, meaning they need 256 / 16 = 16s for the first level. On the second level we have now 128 positions to add => 128 / 16 = 8s , ... When we have < 16 positions at a tree level, it still needs 1s causing some employers to do nothing. Therefore, we can reduce with 8 employers using parallel processing the time from 255s to
+
+     Level:   1 |   2 |  3 |  4 |  5 |  6 |  7 |  8 |  9	
+#positions: 256 | 128 | 64 | 32 | 16 |  8 |  4 |  2 |  1 
+      Time: 16s |  8s | 4s | 2s | 1s | 1s | 1s | 1s |   
+
+      ==> total time: 16 + 8 + 4 + 2 + 4 = 34s
+      ==> with 8 cashiers we can compute the total in only 34s!
+
+ (3) If we have an infinite number of employersm there is still a lower bound how fast we can go when assuming 1s per summation. Given n bags, we need at least d = ceil ( ld (n) ) seconds (ld is the logarithmus dualis, i.e. ld(x) = log(x) / log (2) )
+ Given n = 256s we need at least d = 8s
+
+The reason for this is, that when we want to sum 256 positions we can only use 128 employees concurrently at the same time (as any of these needs 2 bags to perform a summation!). After 128 employees summed 256 positions to 128 positions, now half of them will become workless using the same argument. Thus, when progressing the binary tree at each level we can use less employees which will lead us to the stated lower bound.
+
+ (4)  see P5.png
+
+ (5) As long as we have more positions than employers at a level (or different stated more tasks than nodes) there is no need for communication as we can assign each employer a block of data. Let us therefore define b/2 as the number of employees (each employee can handle 2 bags!). Thus, each employee receives bsize = ceil(n / b) bags (If there aren't any bags left, some employers will take a break or add 0 to treat the special case). After each employee finished summing his block of data  in bsize -1 seconds, we need now to communicate results. Imagining the binary tree, we have now a tree of height d_b = ceil( ld (b) ). Thus in the first level of this tree we have floor(b/2) employers sending their number to floor(b/2) other employers (which will then sum them, if we have an odd number we need to take an extra layer into account which is incorparated in the ceiled height). This will result in (2^d_b - 1) * 1s time for communication on top of the time for computation with d_b = ceil( ld(b) ) the height of the combination tree (if we assume a bus / exclusive communication channel!). When we have a communication channel which allows concurrent communication, we do need only d_b additional seconds for communication.
+
+ In our example here, we have n = 256, b/2 = 8 ==> b = 16. an easy calculation shows bsize = 16, d_b = 4. 
```

Comment:
The question seems unclear as to whether we have 8 or infinite employees in part 4.  Good logic regardless